tests/PyomoTest.py

```diff
@@ -0,0 +1,283 @@
+  m.gen_size = generator_size
+
+  # set up the generator's dispatch variables
+  generator_dispatch = np.zeros((generator_life, hours_in_year), dtype=object)
+  for y in range(generator_life):
+    # make a new pyomo var for this "year"'s dispatch
+    var = pyo.Var(list(range(hours_in_year)), initialize=lambda m, t: 0, bounds=lambda m, t: (0, 100)) #generator_size))
+    setattr(m, f'Gen_disp_year_{y+1}', var)
+    generator_dispatch[y, :] = np.array(list(var.values()))
+
+  generator = build_generator(generator_size, generator_life, generator_dispatch)
+
+  # *** MARKET ***
+  # make the TEAL component
+  market_size = 50 # market depth
+  market_life = 5 # years of operation before replacement
+  # market prices should be interesting, say cos(x) + y/10 where x is hour and y is year
+  prices = np.ones((market_life, hours_in_year))
+  prices[:] *= np.cos(hours) * 100
+  for y in range(market_life):
+    if y % 2 == 0:
+      prices[y, :] *= -1
+
+  # market has similar dispatch to generator, that we'll attach through constraints
+  market_dispatch = np.zeros((market_life, hours_in_year), dtype=object)
+  for y in range(market_life):
+    # make a new pyomo var for this "year"'s dispatch
+    var = pyo.Var(list(range(hours_in_year)), initialize=lambda m, t: 0, bounds=lambda m, t: (0, market_size))
+    setattr(m, f'Market_disp_year_{y+1}', var)
+    market_dispatch[y, :] = np.array(list(var.values()))
+
+  market = build_market(market_size, market_life, prices, market_dispatch)
+
+  # DEBUGG
+
+  metrics = RunCashFlow.run(tealSettings, [generator, market], {}, pyomoVar=True) # Past version was pyomoChk
+
+  # now set up the rest of the Pyomo model
+
+  # *** Constraints ***
+  # hourly conservation of electricity
+  def conserve_elec(y, m, t):
+    gen = getattr(m, f'Gen_disp_year_{y+1}')
+    mkt = getattr(m, f'Market_disp_year_{y+1}')
+    return gen[t] == mkt[t]
+
+  for y in range(project_life):
+    con = pyo.Constraint(m.T, rule=partial(conserve_elec, y))
+    setattr(m, f'conserve_{y+1}', con)
+
+  # hourly dispatch < gen size
+  def size_limit(size, y, m, t):
+    gen = getattr(m, f'Gen_disp_year_{y+1}')
+    return gen[t] <= size
+
+  for y in range(project_life):
+    lim = pyo.Constraint(m.T, rule=partial(size_limit, generator_size, y))
+    setattr(m, f'size_{y+1}', lim)
+
+  # Objective: NPV
+  m.npv = pyo.Objective(expr=metrics['NPV'], sense=pyo.maximize)
+
+  print('DEBUGG PRE:')
+  m.display()
+  # solve pyomo problem
+  solver = pyo.SolverFactory('ipopt')
+  results = solver.solve(m)
+  print('\n\n')
+  print('DEBUGG POST:')
+  m.display()
+  print(m.gen_size.value)
+  calculated = pyo.value(m.npv) # Changed from print(m.npv.value)
+  correct = 10164.740285148146
+  if abs(pyo.value(m.npv) - correct)/correct < 1e-8:
+    print('Success!')
+    sys.exit(0)
+  else:
+    print('ERROR: correct: {:1.3e}, calculated: {:1.3e}, diff {:1.3e}'.format(correct, calculated, correct-calculated))
```

Comment:
printing the error actually won't cause the test to error out; you need a `sys.exit(1)` to register as a failed test

Reply:
Done